README.md

```diff
@@ -60,7 +60,83 @@ Once you have implemented a lower bound search for the start of the range, imple
 
+This code below is a basic binary search:
+
+      def binary_search(arr, x, first, last):
+      
+         # Check base case
+         if last >= first:
+      
+            mid = (last + first) // 2
+      
+            # If element is present at the middle itself
+            if arr[mid] == x:
+                  return mid
+      
+            # If element is smaller than mid, then it can only
+            # be present in left subarray
+            elif arr[mid] > x:
+                  return binary_search(arr, x, first, mid - 1)
+      
+            # Else the element can only be present in right subarray
+            else:
+                  return binary_search(arr, x, mid + 1, last)
+It will return the index where we first find the number we are looking for, even if we have several same numbers.
+
+In our lower_bound binary search, we try to find the index of the number we are looking for which located in the very left, where we assume that the array is already sorted.
+
+      def lower_bound(x, v):
+         def search(arr, v, first, last):
+            if v > arr[last]: # 1
+                  return len(arr)
+
+            if last >= first: # 2
+                  mid = (last + first) // 2
+                  
+                  if mid == 0: # 3
+                     if arr[mid] >= v:
+                        return mid
+                     else:
+                        return mid + 1      
+
+                  if arr[mid] >= v: # 4
+                     # 5
+                     if arr[mid - 1] < v: # 6
+                        return mid
+                     else: # 7
+                        return search(arr, v, first, mid - 1) 
+                  else: # 8
+                     return search(arr, v, mid + 1, last)
+
+         first = 0
+         last = len(x) - 1
+         return search(x, v, first, last)
+
+We will go through the code one by one. I put the number so it can be easier to read.
+
+1. Base case: if the largest number in the array is smaller than integer we are looking for.
+2. We will run the code below if the last index is equal or larger than the first index. 
+3. We end up in this condition if we already reach the very left of array, where last = first = 0. If the value in that index is equal or larger than the number we are looking for, we return 0, otherwise we return 1.
+4. Otherwise, we do the binary search. We check if value in the mid index, is the same or larger than the number we search.
+5. If yes, then we check if the value in the index before the mid is smaller than the number we search. `This part is where it differs from the normal binary search.`
+6. If it also yes, this mid index is indeed the lower bound of the number we are looking for.
+7. Otherwise, we will continue doing a binary search in the  left part of mid index
+8. If the value in mid index we found is smaller than what we are looking for, so it must be in the right side of our mid index.
```

Comment:
This is all correct, but it is a _lot_ more complicated than it needs to be. You are dealing with special cases that would otherwise be handled by the general case, and part of why it gets this complicated is that you want `last` to point _into_ the candidate region instead of one beyond it.

If you make the invariant that for the low/high indices $l,h$ you have:

1.  For all index $i=0, 1, \ldots, l-1: x[i] < v$
2. For all indices $i=h,\ldots n-1: v \leq x[i]$

![lower-bound](https://user-images.githubusercontent.com/3206538/200112462-c9b3f44c-069e-4f8c-8d18-ece02b7dcae7.png)


The lower bound for $v$, the index $i$ where we switch from values smaller than $v$ to those that are equal to or greater, must then be found in $l,l+1,\ldots,h$. It might be a point $j$ where $x[j]=v$ or a point where $x[j]>v$, but in any case, it is in this interval.

If you pick any point $m$ in the interval from $l$ to $h$ (we pick the midpoint for runtime efficiency, not correctness, and any point works for the proof), then we are in one of two cases:

1. $x[m] < v$, which means that $m$ is too low for the switch point. If we move $l$ up to $m+1$, then we will still satisfy the first part of the invariant, and we can continue from there.
2. $x[m] \geq v$, which means that all higher points will be beyond the switch point, and we can move $h$ down to this point and still have the second part of the invariant.

If we reduce the interval size in each iteration (which we will if $m$ is the midpoint and we either move $l$ to $m+1$ or $h$ to $m$), we will eventually end up with $l=h$. Then, because all the values below $l$ are too small, and because none above $h$ are, $l$ is pointing to the lower bound.

That invariant gives you this much shorter and more straightforward lower bound function:

```python
def lower_bound(x: list[int], v: int) -> int:
    lo, hi = 0, len(x)
    while lo < hi:
        m = (lo + hi) // 2
        if x[m] < v:
            lo = m + 1
        else:
            hi = m
    return lo
```